### PR TITLE
Avoid new state creations in the group by reduce step

### PR DESCRIPTION
**Pull request by mfussenegger**
If we're reducing/merging results from the upstreams in the group-by
operation, the incoming rows already contain states. We can use that
existing state instead of creating a new state per first-seen-key.

The effect of this change should be more noticeable the higher the cardinality ratio, the fewer the number of upstreams and the more aggregations involved.

```
V1: 3.1.0-b99678ea93bd4d2a8856544e0b2191d940dafda0
V2: 3.1.0-b95dd1e222b025ed3b975bd58fb4f7915061a9aa

Q: select "cCode", count(*) from uservisits group by "cCode"
C: 15
  4.90% mean difference. Likely NOT significant
             V1    →   V2
  mean:  253.852 → 241.716
  stdev: 272.867 → 233.787
  max:   1476.663 → 1297.000
  min:    55.816 →  48.232
```

### CODE REVIEW

File: sql/src/main/java/io/crate/execution/engine/aggregation/GroupingCollector.java [hunk 1]
```diff
@@ -181,7 +183,26 @@ private GroupingCollector(CollectExpression<Row, ?>[] expressions,
...
+            states = new Object[aggregations.length];
+            statesByKey.put(key, states);
+            for (int i = 0; i < aggregations.length; i++) {
+                states[i] = inputs[i][0].value();
+            }
+        } else {
+            for (int i = 0; i < aggregations.length; i++) {
+                states[i] = aggregations[i].reduce(ramAccountingContext, states[i], inputs[i][0].value());
+            }
+        }
```
Comment (matriv) [inline sql/src/main/java/io/crate/execution/engine/aggregation/GroupingCollector.java]: Shouldn't we add the calculation for the new entry in the map like here: https://github.com/crate/crate/pull/7545/files#diff-2f6039084f6fef0d16e87d1167f7129fR230?

**Comment by matriv** [comment]
I'd also suggest to add a small comment on the `iter` and `reduce` methods.
In the context of this commit/PR it's perfectly clear, but just for someon reading just the code...

**Comment by mfussenegger** [comment]
> I'd also suggest to add a small comment on the iter and reduce methods. 

What do you think is missing / unclear?

File: sql/src/main/java/io/crate/execution/engine/aggregation/GroupingCollector.java (hunk 1 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/execution/engine/aggregation/GroupingCollector.java]: ah right, pushed a fixup

**Comment by matriv** [comment]
@mfussenegger 
for `iter` something like: In the 1st step of aggregations, objects for not already "seen" states must be created.
for `reduce`: During reduce we don't need to create new state objects as those have been already created during the 1 step, see `iter`.

I'm not sure if it makes sense and adds some value...

File: sql/src/main/java/io/crate/execution/engine/aggregation/GroupingCollector.java [hunk 2]
```diff
@@ -181,7 +183,29 @@ private GroupingCollector(CollectExpression<Row, ?>[] expressions,
...
+            expression.setNextRow(row);
+        }
+        K key = keyExtractor.apply(row);
+        Object[] states = statesByKey.get(key);
+        if (states == null) {
+            states = new Object[aggregations.length];
+            for (int i = 0; i < aggregations.length; i++) {
+                states[i] = inputs[i][0].value();
+            }
+            // key size + 32 bytes for entry + 4 bytes for increased capacity
```
Comment (matriv) [inline sql/src/main/java/io/crate/execution/engine/aggregation/GroupingCollector.java]: If doesn't result in bloated up method, maybe extract it?

**Comment by mfussenegger** [comment]
Not sure if such comments make things clearer. They might make sense in the context of this change, under the assumption that things were not optimal before. But if you read this variant from scratch the comments are just repeating what the code says. (`what` vs `why` again)

The question to me is more if the concept of "reduce" vs "iter" is clear - and this implication that "reduce" means the incoming rows contain pre-aggregated results. But in this case I wonder *where* we put the comments to clarify that. Probably in the `AggregationMode` enum?

**Comment by matriv** [comment]
I agree. We shouldn't add more comments if they end up being more confusing or not providing any real extra info.

Adding some general comments to the `AggregateMode` about the phases would be nice, but not in the scope of this PR.

thx!

**Pull request merged**